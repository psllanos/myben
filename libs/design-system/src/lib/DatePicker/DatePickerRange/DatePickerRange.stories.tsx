import type { Story, Meta } from '@storybook/react'
import type { DateRange } from '../selectableRanges'
import { useState } from 'react'
import { DateTime } from 'luxon'
import { DatePickerRange } from './DatePickerRange'

export default {
    title: 'Components/DatePicker/DatePickerRange',
    component: DatePickerRange,
    argTypes: {
        placeholder: { description: 'The Input placeholder value' },
        value: { description: 'Valid ISO 8601 string' },
        minDate: { control: 'text', description: 'Valid ISO 8601 string' },
        maxDate: { control: 'text', description: 'Valid ISO 8601 string' },
    },
} as Meta

export const Base: Story = (args) => {
    const [range, setRange] = useState<Partial<DateRange> | undefined>(undefined)

    return (
        <div className="flex justify-center h-[600px]">
            <DatePickerRange
                {...args}
                value={range}
                selectableRanges={[
                    'this-month',
                    'prior-month',
                    'this-year',
                    'prior-year',
                    {
                        label: 'All time',
                        labelShort: 'All',
                        start: DateTime.now().minus({ years: 3 }).toISODate(), // this would be dynamic in real app
                        end: DateTime.now().toISODate(),
                    },
                ]}
                onChange={setRange}
            />
        </div>
    )
}

export const Tabs: Story = (args) => {
    // const [range, setRange] = useState<Partial<DateRange> | undefined>(undefined)

    const [range, setRange] = useState<Partial<DateRange> | undefined>({
        start: '2022-02-01',
        end: '2022-02-28',
    })

    return (
        <div className="flex justify-center h-[600px]">
            <DatePickerRange
                {...args}
                variant="tabs"
                value={range}
                selectableRanges={[
                    'day',
                    'last-7-days',
                    'last-30-days',
                    'last-90-days',
                    {
                        label: 'All time',
                        labelShort: 'All',
                        start: DateTime.now().minus({ years: 3 }).toISODate(), // this would be dynamic in real app
                        end: DateTime.now().toISODate(),
                    },
                ]}
                onChange={setRange}
                popperPlacement="bottom-end"
            />
        </div>
    )
}

export const TabsWithCustomRange: Story = (args) => {
    // const [range, setRange] = useState<Partial<DateRange> | undefined>(undefined)

    const [range, setRange] = useState<Partial<DateRange> | undefined>({
        start: '2022-02-01',
        end: '2022-02-28',
    })

    return (
        <div className="flex justify-center h-[600px]">
            <DatePickerRange
                {...args}
                variant="tabs-custom"
                value={range}
                selectableRanges={[
                    'day',
                    'last-7-days',
                    'last-30-days',
                    'last-90-days',
                    {
                        label: 'All time',
                        labelShort: 'All',
                        start: DateTime.now().minus({ years: 3 }).toISODate(), // this would be dynamic in real app
                        end: DateTime.now().toISODate(),
                    },
                ]}
                onChange={setRange}
                popperPlacement="bottom-end"
            />
        </div>
    )
}
