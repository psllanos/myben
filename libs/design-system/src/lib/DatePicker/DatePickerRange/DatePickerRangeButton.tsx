import type { DateRange } from '../selectableRanges'
import { Popover } from '@headlessui/react'
import { DateTime } from 'luxon'
import { RiCalendarEventFill } from 'react-icons/ri'

export interface DatePickerRangeButtonProps {
    value?: Partial<DateRange>
    setPopperReferenceElement: (el: HTMLButtonElement) => void
}

// The datepicker button that toggles the panel to open/close, and displays the currently selected date range
export function DatePickerRangeButton({
    value,
    setPopperReferenceElement,
}: DatePickerRangeButtonProps) {
    return (
        <Popover.Button ref={setPopperReferenceElement} data-testid="datepicker-range-toggle-icon">
            {({ open }) => {
                return (
                    <div
                        className="flex items-center justify-between font-normal text-gray-25 text-base border border-gray-200 rounded py-2 px-4"
                        data-testid="date-range"
                    >
                        {open || !value ? (
                            <span>Select a date range</span>
                        ) : (
                            <>
                                {value.start && (
                                    <span>
                                        {DateTime.fromISO(value.start).toFormat('MMM dd, yyyy')}
                                    </span>
                                )}
                                <span className="mx-2 text-gray-200">to</span>
                                {value.end && (
                                    <span>
                                        {DateTime.fromISO(value.end).toFormat('MMM dd, yyyy')}
                                    </span>
                                )}
                            </>
                        )}
                        <RiCalendarEventFill className="ml-4 text-lg" />
                    </div>
                )
            }}
        </Popover.Button>
    )
}
