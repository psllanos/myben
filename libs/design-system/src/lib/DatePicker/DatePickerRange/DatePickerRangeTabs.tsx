import type { Dispatch, SetStateAction } from 'react'
import type { DateRange, SelectableDateRange } from '..'
import { useMemo } from 'react'
import { Popover, Tab } from '@headlessui/react'
import classNames from 'classnames'
import { RiCalendarEventFill } from 'react-icons/ri'

export type DatePickerRangeTabsProps =
    | {
          variant: 'custom'
          value?: Partial<DateRange>
          onChange: (range: DateRange) => void
          tabs: Array<SelectableDateRange | 'custom'>
          setReferenceElement?: Dispatch<SetStateAction<HTMLElement | null | undefined>>
      }
    | {
          variant: 'default'
          value?: Partial<DateRange>
          onChange: (range: DateRange) => void
          tabs: SelectableDateRange[]
          setReferenceElement?: never
      }

function DefaultTab({ value, selected }: { value: string; selected: boolean }) {
    return (
        <span
            className={classNames(
                'flex items-center px-4 py-1 text-base rounded hover:bg-gray-300',
                selected ? 'bg-gray-400 text-white shadow' : 'text-gray-100'
            )}
        >
            {value}
        </span>
    )
}

const NOT_SELECTED_TAB_INDEX = 999

// The tabs that persistently show and can toggle the panel to open/close
export function DatePickerRangeTabs(props: DatePickerRangeTabsProps) {
    const selectedIndex = useMemo(() => {
        // No range, no tab selected
        if (!props.value?.start || !props.value?.end) {
            return NOT_SELECTED_TAB_INDEX
        }

        const tabIndex = props.tabs.findIndex(
            (tab) =>
                typeof tab !== 'string' && // not custom tab
                tab.start === props.value?.start &&
                tab.end === props.value?.end
        )

        // known range, select tab
        if (tabIndex !== -1) {
            return tabIndex
        }

        const customTabIndex = props.tabs.findIndex(
            (tab) => typeof tab === 'string' && tab === 'custom'
        )

        // custom range, select custom tab
        if (customTabIndex !== -1) {
            return customTabIndex
        }

        return NOT_SELECTED_TAB_INDEX
    }, [props.tabs, props.value?.start, props.value?.end])

    // Render the custom variant (a datepicker shows up as the very last tab)
    if (props.variant === 'custom') {
        return (
            <div className="flex items-center">
                <Tab.Group
                    selectedIndex={selectedIndex}
                    onChange={(index: number) => {
                        const tab = props.tabs[index]

                        // If the "custom" tab is clicked, we defer state updates until the user interacts with the calendar panel and submits
                        if (tab !== 'custom') {
                            props.onChange({ start: tab.start, end: tab.end })
                        }
                    }}
                >
                    <Tab.List className="inline-flex items-center p-1 gap-x-2 text-white">
                        {props.tabs.map((tab) => (
                            <Tab key={typeof tab === 'string' ? tab : tab.label}>
                                {({ selected }) =>
                                    tab === 'custom' ? (
                                        <Popover.Button
                                            ref={props.setReferenceElement}
                                            as="div" // Cannot render a button inside of a button (<Tab> renders as a button by default)
                                            data-testid="datepicker-range-toggle-icon-tabs"
                                            className={classNames(
                                                'px-4 flex items-center text-gray-100 rounded h-8 hover:bg-gray-300',
                                                selected && 'bg-gray-400 text-white shadow'
                                            )}
                                        >
                                            <RiCalendarEventFill
                                                className={classNames(
                                                    'text-lg cursor-pointer',
                                                    selected && 'text-white'
                                                )}
                                            />
                                        </Popover.Button>
                                    ) : (
                                        <DefaultTab value={tab.labelShort} selected={selected} />
                                    )
                                }
                            </Tab>
                        ))}
                    </Tab.List>
                </Tab.Group>
            </div>
        )
    }

    // Default implementation - no datepicker is available to click here
    return (
        <div className="flex items-center">
            <Tab.Group
                selectedIndex={selectedIndex}
                onChange={(index: number) => {
                    const { start, end } = props.tabs[index]
                    props.onChange({ start, end })
                }}
            >
                <Tab.List className="inline-flex items-center p-1 text-white">
                    {props.tabs.map((tab) => (
                        <Tab key={tab.labelShort}>
                            {({ selected }) => (
                                <DefaultTab value={tab.labelShort} selected={selected} />
                            )}
                        </Tab>
                    ))}
                </Tab.List>
            </Tab.Group>
        </div>
    )
}
