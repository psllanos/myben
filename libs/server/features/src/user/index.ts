export * from './user.processor'
export * from './user.service'
export * from './onboarding.schema'

export type { OnboardingState, RegisteredStep } from './onboarding.service'
