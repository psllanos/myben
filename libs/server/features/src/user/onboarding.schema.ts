import { z } from 'zod'

export const UpdateOnboardingSchema = z.discriminatedUnion('flow', [
    z.object({
        flow: z.literal('main'),
        updates: z
            .object({
                key: z.enum([
                    'intro',
                    'profile',
                    'verifyEmail',
                    'firstAccount',
                    'accountSelection',
                    'terms',
                    'maybe',
                    'welcome',
                ]),
                markedComplete: z.boolean(),
            })
            .array(),
    }),
    z.object({
        flow: z.literal('sidebar'),
        markedComplete: z.boolean().optional(),
        updates: z
            .object({
                key: z.enum([
                    'connect-depository',
                    'connect-investment',
                    'connect-liability',
                    'add-crypto',
                    'add-property',
                    'add-vehicle',
                    'add-other',
                    'upgrade-account',
                    'create-plan',
                ]),
                markedComplete: z.boolean(),
            })
            .array(),
    }),
])
