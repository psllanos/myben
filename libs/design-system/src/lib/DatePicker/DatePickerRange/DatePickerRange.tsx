import type * as PopperJs from '@popperjs/core'
import type { DateRange, SelectableDateRange, SelectableRangeKeys } from '../selectableRanges'

import { Popover, Portal } from '@headlessui/react'
import classNames from 'classnames'
import { DateTime } from 'luxon'
import { RiArrowRightLine } from 'react-icons/ri'
import { useEffect, useMemo, useState } from 'react'
import { usePopper } from 'react-popper'
import { Button } from '../../Button'
import { DatePickerRangeCalendar } from './DatePickerRangeCalendar'
import { DatePickerInput } from '../DatePickerInput'
import { getNormalizedRanges } from '../selectableRanges'
import { DatePickerQuickSelect } from '../DatePickerQuickSelect'
import { DatePickerRangeButton } from './DatePickerRangeButton'
import { DatePickerRangeTabs } from './DatePickerRangeTabs'

export interface DatePickerRangeProps {
    variant?: 'default' | 'tabs' | 'tabs-custom'
    value?: Partial<DateRange>
    selectableRanges: Array<SelectableRangeKeys | SelectableDateRange>
    onChange: (range: DateRange) => void
    className?: string
    minDate?: string
    maxDate?: string
    popperPlacement?: PopperJs.Placement
    popperStrategy?: PopperJs.PositioningStrategy
}

export function DatePickerRange({
    variant = 'default',
    value,
    selectableRanges,
    onChange,
    className,
    minDate = DateTime.now().minus({ years: 50 }).toISODate(),
    maxDate = DateTime.now().toISODate(),
    popperPlacement = 'bottom-end',
    popperStrategy = 'fixed',
}: DatePickerRangeProps) {
    // PopperJS configuration: positions the datepicker panel appropriately based on screen size and parent elements
    const [referenceElement, setReferenceElement] = useState<HTMLElement | null>()
    const [popperElement, setPopperElement] = useState<HTMLDivElement | null>()
    const { styles, attributes } = usePopper(referenceElement, popperElement, {
        placement: popperPlacement,
        strategy: popperStrategy,
        modifiers: [
            {
                name: 'offset',
                options: {
                    offset: [0, variant === 'default' ? 8 : 16],
                },
            },
            {
                name: 'preventOverflow',
                options: {
                    altAxis: true,
                },
            },
        ],
    })

    const [range, setRange] = useState(value)
    const [rangeError, setRangeError] = useState(false)
    const [startError, setStartError] = useState(false)
    const [endError, setEndError] = useState(false)

    useEffect(() => {
        setRangeError(
            range && !!range.start && !!range.end
                ? DateTime.fromISO(range.start) > DateTime.fromISO(range.end)
                : false
        )
    }, [range])

    const tabs = useMemo<SelectableDateRange[]>(() => {
        return getNormalizedRanges(selectableRanges)
    }, [selectableRanges])

    return (
        <Popover className={classNames(className, 'relative z-10')}>
            {variant === 'tabs' ? (
                <DatePickerRangeTabs
                    variant="default"
                    tabs={tabs}
                    value={value}
                    onChange={(range) => {
                        onChange(range)
                        setRange(range)
                    }}
                />
            ) : variant === 'tabs-custom' ? (
                <DatePickerRangeTabs
                    variant="custom"
                    tabs={[...tabs, 'custom']}
                    value={value}
                    onChange={(range) => {
                        onChange(range)
                        setRange(range)
                    }}
                    setReferenceElement={setReferenceElement}
                />
            ) : (
                <DatePickerRangeButton
                    value={value}
                    setPopperReferenceElement={setReferenceElement}
                />
            )}

            <Portal>
                <Popover.Panel
                    className={classNames(
                        'border border-gray-500 rounded bg-gray-700 shadow-lg z-50'
                    )}
                    ref={setPopperElement}
                    style={styles.popper}
                    data-testid="datepicker-range-panel"
                    {...attributes.popper}
                >
                    {({ close }) => (
                        <div
                            className={classNames('flex gap-6 p-4 rounded text-gray-25 text-base')}
                        >
                            {variant === 'default' && (
                                <div className="hidden sm:block whitespace-nowrap">
                                    <DatePickerQuickSelect
                                        ranges={tabs}
                                        value={range}
                                        onChange={setRange}
                                    />
                                </div>
                            )}
                            <div className="flex flex-col">
                                <div className="grow">
                                    <DatePickerRangeCalendar
                                        range={range}
                                        onChange={setRange}
                                        minDate={minDate}
                                        maxDate={maxDate}
                                        rangeInputs={
                                            <div className="flex flex-col xs:flex-row justify-between gap-2 h-18 my-2">
                                                <DatePickerInput
                                                    value={range ? range.start : undefined}
                                                    onChange={(start: string) =>
                                                        setRange((previous) => ({
                                                            ...previous,
                                                            start,
                                                        }))
                                                    }
                                                    hasError={startError || rangeError}
                                                    onError={(err) => setStartError(!!err)}
                                                    minDate={minDate}
                                                    maxDate={maxDate}
                                                    className="w-full xs:w-36"
                                                />
                                                <span className="flex justify-center items-center text-gray-50">
                                                    <RiArrowRightLine size={20} />
                                                </span>
                                                <DatePickerInput
                                                    value={range ? range.end : undefined}
                                                    onChange={(end: string) =>
                                                        setRange((previous) => ({
                                                            ...previous,
                                                            end,
                                                        }))
                                                    }
                                                    hasError={endError || rangeError}
                                                    onError={(err) => setEndError(!!err)}
                                                    minDate={minDate}
                                                    maxDate={maxDate}
                                                    className="w-full mb-4 xs:mb-0 xs:w-36" // On mobile, calendar sits under the inputs
                                                />
                                            </div>
                                        }
                                        controlButtons={
                                            <div className="flex items-center justify-end">
                                                <Button
                                                    className="mr-4"
                                                    variant="secondary"
                                                    onClick={() => close()}
                                                >
                                                    Cancel
                                                </Button>
                                                <Button
                                                    onClick={() => {
                                                        onChange({
                                                            start: range!.start!,
                                                            end: range!.end!,
                                                        })
                                                        close()
                                                    }}
                                                    disabled={
                                                        rangeError ||
                                                        startError ||
                                                        endError ||
                                                        !range ||
                                                        !range.start ||
                                                        !range.end
                                                    }
                                                >
                                                    Apply
                                                </Button>
                                            </div>
                                        }
                                    />
                                </div>
                            </div>
                        </div>
                    )}
                </Popover.Panel>
            </Portal>
        </Popover>
    )
}
