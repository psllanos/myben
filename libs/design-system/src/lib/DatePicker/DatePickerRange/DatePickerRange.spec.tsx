import user from '@testing-library/user-event'
import { fireEvent, getQueriesForElement, render, screen, waitFor } from '@testing-library/react'
import { DatePickerRange } from './DatePickerRange'
import { DateTime } from 'luxon'

// Set date to Oct 29, 2021 to keep snapshots consistent
beforeAll(() => jest.useFakeTimers().setSystemTime(new Date('2021-10-29 12:00:00')))

// DatePicker configuration
const minDate = DateTime.now().minus({ years: 2 })
const maxDate = DateTime.now()

describe('<DatePickerRange />', () => {
    describe('When datepicker is closed', () => {
        it('should have placeholder when empty', () => {
            const onChangeMock = jest.fn()

            const component = render(
                <DatePickerRange
                    selectableRanges={['this-month', 'prior-month']}
                    minDate={minDate.toISODate()}
                    maxDate={maxDate.toISODate()}
                    value={undefined}
                    onChange={onChangeMock}
                />
            )

            expect(screen.queryByText('Select a date range')).toBeInTheDocument()
            expect(screen.queryByText('Cancel')).not.toBeInTheDocument()
            expect(screen.queryByText('Apply')).not.toBeInTheDocument()
            expect(component).toMatchSnapshot()
        })

        it('should have date when valid', () => {
            const onChangeMock = jest.fn()

            const component = render(
                <DatePickerRange
                    selectableRanges={['this-month', 'prior-month']}
                    minDate={'2021-01-01'}
                    maxDate={'2022-01-01'}
                    value={{ start: '2021-01-01', end: '2022-01-01' }}
                    onChange={onChangeMock}
                />
            )

            expect(screen.queryByText('Jan 01, 2021')).toBeInTheDocument()
            expect(screen.queryByText('to')).toBeInTheDocument()
            expect(screen.queryByText('Jan 01, 2022')).toBeInTheDocument()
            expect(screen.queryByText('Cancel')).not.toBeInTheDocument()
            expect(screen.queryByText('Apply')).not.toBeInTheDocument()
            expect(component).toMatchSnapshot()
        })
    })

    describe('When datepicker is expanded', () => {
        it('should be able to select a date from the dropdown while clicking', async () => {
            const onChangeMock = jest.fn()

            const start = DateTime.fromISO('2021-12-01')
            const end = DateTime.fromISO('2022-01-31')

            render(
                <DatePickerRange
                    selectableRanges={['day', 'last-30-days']}
                    minDate={start.toISODate()}
                    maxDate={end.toISODate()}
                    value={{ start: start.toISODate(), end: end.toISODate() }}
                    onChange={onChangeMock}
                />
            )

            // Open the modal
            fireEvent.click(screen.getByTestId('datepicker-range-toggle-icon'))

            await waitFor(() =>
                expect(screen.getByTestId('datepicker-range-panel')).toBeInTheDocument()
            )

            expect(screen.queryByText('Cancel')).toBeInTheDocument()
            expect(screen.queryByText('Apply')).toBeInTheDocument()

            expect(screen.queryByText(end.monthShort)).toBeInTheDocument()
            expect(screen.queryByText(end.year)).toBeInTheDocument()

            // Go back a month
            fireEvent.click(screen.getByTestId('datepicker-range-back-arrow'))
            expect(screen.queryByText(start.monthShort)).toBeInTheDocument()
            expect(screen.queryByText(start.year)).toBeInTheDocument()

            // Select start date
            fireEvent.click(
                await getQueriesForElement(screen.getByTestId('day-cells')).findByText('1')
            )

            // Go forward a month
            fireEvent.click(screen.getByTestId('datepicker-range-next-arrow'))
            expect(screen.queryByText(end.monthShort)).toBeInTheDocument()
            expect(screen.queryByText(end.year)).toBeInTheDocument()

            // Select end date
            fireEvent.click(
                await getQueriesForElement(screen.getByTestId('day-cells')).findByText('31')
            )

            // Submit
            fireEvent.click(screen.getByText('Apply'))

            expect(onChangeMock).toHaveBeenCalledWith({
                start: start.toISODate(),
                end: end.toISODate(),
            })

            await waitFor(() =>
                expect(screen.queryByTestId('datepicker-range-panel')).not.toBeInTheDocument()
            )

            expect(screen.queryByText('Cancel')).not.toBeInTheDocument()
            expect(screen.queryByText('Apply')).not.toBeInTheDocument()
        })
        it('should be able to select a date from the dropdown using inputs', async () => {
            const onChangeMock = jest.fn()

            const start = DateTime.fromISO('2022-01-01')
            const end = DateTime.fromISO('2022-01-20')

            render(
                <DatePickerRange
                    selectableRanges={['day', 'last-30-days']}
                    minDate={start.toISODate()}
                    maxDate={end.toISODate()}
                    value={{ start: start.toISODate(), end: end.toISODate() }}
                    onChange={onChangeMock}
                />
            )

            // Open the modal
            fireEvent.click(screen.getByTestId('datepicker-range-toggle-icon'))

            await waitFor(() =>
                expect(screen.getByTestId('datepicker-range-panel')).toBeInTheDocument()
            )

            expect(screen.queryByText('Cancel')).toBeInTheDocument()
            expect(screen.queryByText('Apply')).toBeInTheDocument()

            const dateInputs = screen.getAllByRole('textbox')

            user.type(dateInputs[0], start.toFormat('MMddyyyy'))
            user.type(dateInputs[1], end.toFormat('MMddyyyy'))

            expect(screen.queryByText(end.monthShort)).toBeInTheDocument()
            expect(screen.queryByText(end.year)).toBeInTheDocument()

            // Submit
            fireEvent.click(screen.getByText('Apply'))

            expect(onChangeMock).toHaveBeenCalledWith({
                start: start.toISODate(),
                end: end.toISODate(),
            })

            await waitFor(() =>
                expect(screen.queryByTestId('datepicker-range-panel')).not.toBeInTheDocument()
            )

            expect(screen.queryByText('Cancel')).not.toBeInTheDocument()
            expect(screen.queryByText('Apply')).not.toBeInTheDocument()
        })

        it('should be able to select a date using the quick select sidebar', async () => {
            const onChangeMock = jest.fn()

            render(
                <DatePickerRange
                    selectableRanges={['prior-month', 'this-month']}
                    onChange={onChangeMock}
                />
            )

            // Open the modal
            fireEvent.click(screen.getByTestId('datepicker-range-toggle-icon'))

            await waitFor(() =>
                expect(screen.getByTestId('datepicker-range-panel')).toBeInTheDocument()
            )

            expect(screen.queryByText('Cancel')).toBeInTheDocument()
            expect(screen.queryByText('Apply')).toBeInTheDocument()

            fireEvent.click(screen.getByText('Last month'))

            const lastMonth = DateTime.now().minus({ months: 1 })

            expect(screen.queryByText(lastMonth.monthShort)).toBeInTheDocument()
            expect(screen.queryByText(lastMonth.year)).toBeInTheDocument()

            // Submit
            fireEvent.click(screen.getByText('Apply'))

            expect(onChangeMock).toHaveBeenCalledWith({
                start: lastMonth.startOf('month').toISODate(),
                end: lastMonth.endOf('month').toISODate(),
            })

            await waitFor(() =>
                expect(screen.queryByTestId('datepicker-range-panel')).not.toBeInTheDocument()
            )

            expect(screen.queryByText('Cancel')).not.toBeInTheDocument()
            expect(screen.queryByText('Apply')).not.toBeInTheDocument()
        })
    })
})
