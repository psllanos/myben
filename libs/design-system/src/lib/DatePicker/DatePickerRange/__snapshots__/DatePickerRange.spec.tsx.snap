// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePickerRange /> When datepicker is closed should have date when valid 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="headlessui-portal-root"
    >
      <div
        data-headlessui-portal=""
      />
      <div
        data-headlessui-portal=""
      />
    </div>
    <div>
      <div
        class="relative z-10"
        data-headlessui-state=""
      >
        <button
          aria-expanded="false"
          data-headlessui-state=""
          data-testid="datepicker-range-toggle-icon"
          id="headlessui-popover-button-:r5:"
          type="button"
        >
          <div
            class="flex items-center justify-between font-normal text-gray-25 text-base border border-gray-200 rounded py-2 px-4"
            data-testid="date-range"
          >
            <span>
              Jan 01, 2021
            </span>
            <span
              class="mx-2 text-gray-200"
            >
              to
            </span>
            <span>
              Jan 01, 2022
            </span>
            <svg
              class="ml-4 text-lg"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M17 3h4a1 1 0 0 1 1 1v16a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h4V1h2v2h6V1h2v2zM4 9v10h16V9H4zm2 4h5v4H6v-4z"
                />
              </g>
            </svg>
          </div>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative z-10"
      data-headlessui-state=""
    >
      <button
        aria-expanded="false"
        data-headlessui-state=""
        data-testid="datepicker-range-toggle-icon"
        id="headlessui-popover-button-:r5:"
        type="button"
      >
        <div
          class="flex items-center justify-between font-normal text-gray-25 text-base border border-gray-200 rounded py-2 px-4"
          data-testid="date-range"
        >
          <span>
            Jan 01, 2021
          </span>
          <span
            class="mx-2 text-gray-200"
          >
            to
          </span>
          <span>
            Jan 01, 2022
          </span>
          <svg
            class="ml-4 text-lg"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17 3h4a1 1 0 0 1 1 1v16a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h4V1h2v2h6V1h2v2zM4 9v10h16V9H4zm2 4h5v4H6v-4z"
              />
            </g>
          </svg>
        </div>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<DatePickerRange /> When datepicker is closed should have placeholder when empty 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative z-10"
        data-headlessui-state=""
      >
        <button
          aria-expanded="false"
          data-headlessui-state=""
          data-testid="datepicker-range-toggle-icon"
          id="headlessui-popover-button-:r0:"
          type="button"
        >
          <div
            class="flex items-center justify-between font-normal text-gray-25 text-base border border-gray-200 rounded py-2 px-4"
            data-testid="date-range"
          >
            <span>
              Select a date range
            </span>
            <svg
              class="ml-4 text-lg"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M17 3h4a1 1 0 0 1 1 1v16a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h4V1h2v2h6V1h2v2zM4 9v10h16V9H4zm2 4h5v4H6v-4z"
                />
              </g>
            </svg>
          </div>
        </button>
      </div>
    </div>
    <div
      id="headlessui-portal-root"
    >
      <div
        data-headlessui-portal=""
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="relative z-10"
      data-headlessui-state=""
    >
      <button
        aria-expanded="false"
        data-headlessui-state=""
        data-testid="datepicker-range-toggle-icon"
        id="headlessui-popover-button-:r0:"
        type="button"
      >
        <div
          class="flex items-center justify-between font-normal text-gray-25 text-base border border-gray-200 rounded py-2 px-4"
          data-testid="date-range"
        >
          <span>
            Select a date range
          </span>
          <svg
            class="ml-4 text-lg"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17 3h4a1 1 0 0 1 1 1v16a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h4V1h2v2h6V1h2v2zM4 9v10h16V9H4zm2 4h5v4H6v-4z"
              />
            </g>
          </svg>
        </div>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
